modules/uhsg_sisu/src/Services/StudyRightsService.php

```diff
@@ -0,0 +1,368 @@
+
+    // Handle specialisation properly
+    return $this->degreeProgramWithSpecialisation($degreeprogram, $degreeprogramchild);
+  }
+
+  /**
+   * Get Primary Study Right
+   */
+  private function getPrimaryStudyRight($studyRightPrimalityChain, $studyRights) {
+    // Make sure we have all the needed data.
+    if (!$studyRightPrimalityChain || !$studyRightPrimalityChain['studyRightPrimalities']
+      || !strlen($studyRightPrimalityChain['studyRightPrimalities']) || !$studyRights || !strlen($studyRights)) {
+      return null;
+    }
+
+    // Make sure all data looks ok and return proper id if we have active studyright
+    $studyrightprimalities = $studyRightPrimalityChain['studyRightPrimalities'];
+
+    // Loop trough all studyrightprimalities and find "last" active primality
+    foreach($studyrightprimalities as $id => $studyrightprimality) {
+      if($studyrightprimality['startDate'] && !$studyrightprimality['endDate'] && $studyrightprimality['documentState'] == 'ACTIVE') {
+        $studyRightId = $studyrightprimality['studyRightId'];
+      }
+    }
+
+    // Loop trough studyrights and return the correct one
+    foreach($studyrights as $id => $studyright) {
+      if($studyright['id'] == $studyRightId) {
+        return $studyright;
+      }
+    }
+  }
+
+  /*
+  https://jira.it.helsinki.fi/browse/OPISKELU-506
+  Students of the Faculty of Educational Sciences get special treatment.
+  1) Their specialisation name is added to the name of their degree programme.
+  2) The Guide News are fetched for them using a combination of the degree programme and specialisation codes.
+  An additional twist: the specialisation codes are different in Sisu and Oodi. Guide news API currently only supports Oodi specialisation codes.
+  Since we get the Sisu specialisation module groupId from the study rights, and we have to fetch news using the Oodi code,
+  we need to have a mapping from Sisu module groupId to Oodi specialisation codes.
+  Note: the Sisu module ids in question are the same in QA and production.
+  */
+  private function degreeProgramWithSpecialisation($degreeProgram, $specialisation) {
+    if ($degreeProgram && $specialisation && $specialisation['groupId']) {
+      // Read file
+      $path = getcwd() . "/". drupal_get_path('module', 'uhsg_sisu') . "/src/Services/sisu-oodi-codes.json";
+      $sisu_oodi_codes = Json::decode(file_get_contents($path));
+
+      // Traverse trough all the oodi-sisu mappings.
+      foreach($sisu_oodi_codes as $group) {
+        // If we encounter a mapping that fits, then mark that
+        if($group['groupId'] == $specialisation['groupId']) {
+          $oodiMapping = $group;
+        }
+      }
+
+      // If we have a match, then we need to modify our degreeprogram before returning it.
+      if ($oodiMapping) {
+        $degreeProgram['code'] .= $oodiMapping['oodiSpecialisationCode'];
```

Comment:
This would be an example where I think an !empty()-heck could be useful (as well as many other places in the PR).

Then again you've added a lot of error handling so I guess its ok as is.